gusto/linear_solvers.py

```diff
@@ -220,6 +220,182 @@ def solve(self):
+    (3) Reconstruct theta
+
+    :arg state: a :class:`.State` object containing everything else.
+    :arg quadrature degree: tuple (q_h, q_v) where q_h is the required
+    quadrature degree in the horizontal direction and q_v is that in
+    the vertical direction
+    :arg params (optional): solver parameters
+    """
+
+    def __init__(self, state, quadrature_degree=None, params=None):
+
+        self.state = state
+
+        if quadrature_degree is not None:
+            self.quadrature_degree = quadrature_degree
+        else:
+            dgspace = state.spaces("DG")
+            if any(deg > 2 for deg in dgspace.ufl_element().degree()):
+                warning("default quadrature degree most likely not sufficient for this degree element")
+            self.quadrature_degree = (5, 5)
+
+        if params is None:
+            self.params = {'pc_type': 'fieldsplit',
+                           'pc_fieldsplit_type': 'schur',
+                           'ksp_type': 'gmres',
+                           'ksp_max_it': 100,
+                           'ksp_gmres_restart': 50,
+                           'pc_fieldsplit_schur_fact_type': 'FULL',
+                           'pc_fieldsplit_schur_precondition': 'selfp',
+                           'fieldsplit_0_ksp_type': 'preonly',
+                           'fieldsplit_0_pc_type': 'bjacobi',
+                           'fieldsplit_0_sub_pc_type': 'ilu',
+                           'fieldsplit_1_ksp_type': 'preonly',
+                           'fieldsplit_1_pc_type': 'gamg',
+                           'fieldsplit_1_mg_levels_ksp_type': 'chebyshev',
+                           'fieldsplit_1_mg_levels_ksp_chebyshev_estimate_eigenvalues': True,
+                           'fieldsplit_1_mg_levels_ksp_chebyshev_estimate_eigenvalues_random': True,
+                           'fieldsplit_1_mg_levels_ksp_max_it': 1,
+                           'fieldsplit_1_mg_levels_pc_type': 'bjacobi',
+                           'fieldsplit_1_mg_levels_sub_pc_type': 'ilu'}
+        else:
+            self.params = params
+
+        # setup the solver
+        self._setup_solver()
+
+    def _setup_solver(self):
+        state = self.state      # just cutting down line length a bit
+        dt = state.timestepping.dt
+        beta = dt*state.timestepping.alpha
+        cp = state.parameters.cp
+        mu = state.mu
+        Vu = state.spaces("HDiv")
+        Vu_broken = FunctionSpace(BrokenElement(Vu.ufl_element()))
+        Vtheta = state.spaces("HDiv_v")
+        Vrho = state.spaces("DG")
+        HOW DO WE GET THE TRACE SPACE IN THE EXTRUDED CASE?
+        Vtrace = FunctionSpace(mesh, "HDiv Trace", tdegree)
+
+        # Split up the rhs vector (symbolically)
+        u_in, rho_in, theta_in = split(state.xrhs)
+
+        # Build the reduced function space for u,rho
+        M = MixedFunctionSpace((Vu_broken, Vrho, Vtrace))
+        w, phi, dl = TestFunctions(M)
+        u, rho, l0 = TrialFunctions(M)
+
```

Comment:
Ah so you're defining a triple `(Vu_broken, Vrho, Vtrace)`. Due to the whole "pyop2 can't write into mixed data objects," you'll want to define the trace space separately and just leave `M = MixedFunctionSpace((Vu_broken, Vrho))`. So:

```python
M = MixedFunctionSpace((Vu_broken, Vrho))
w, phi = TestFunctions(M)
u, rho = TrialFunctions(M)
l0 = TrialFunction(Vtrace)
dl = TestFunction(Vtrace)
```

This is because computing the schur complement for the multiplier system will result in an expression whose arguments lie in a non-mixed space.